OpenXR backend implementation

morallo opened the pull request:
Draft PR to discuss and follow-up process

morallo committed:
OpenXR configuration options and initialization

morallo committed:
OpenXR configuration options and initialization

morallo committed:
Implement buffer copy to swapchain. No head tracking.

morallo committed:
Implement buffer copy to swapchain. No head tracking.

morallo committed:
Only initialize OpenXR when configured and available.

morallo committed:
Only initialize OpenXR when configured and available.

Prof-Butts commented on impl11/ddraw/VRRendererOpenXR.cpp:
```diff
@@ -0,0 +1,139 @@
+bool VRRendererOpenXR::is_available()
+{
+	// OpenXR will fail to initialize if we ask for an extension that OpenXR
+	// can't provide! So we need to check our all extensions before 
```
Did you mean "check out all..."?

Prof-Butts commented on impl11/ddraw/VRRendererOpenXR.cpp:
```diff
@@ -0,0 +1,139 @@
+
+	log_debug("[DBG][OpenXR] Extensions available");
+	for (size_t i = 0; i < xr_exts.size(); i++) {
+		log_debug("- %s\n", xr_exts[i].extensionName);
```
I think you need to add the "[DBG]" in this log_debug call, also: there's no need for the "\n" at the end of the line

morallo replied:
Indeed. Fixed.

Prof-Butts commented on impl11/ddraw/DeviceResources.cpp:
```diff
@@ -1502,7 +1504,7 @@ HRESULT DeviceResources::OnSizeChanged(HWND hWnd, DWORD dwWidth, DWORD dwHeight)
 			sd.SwapEffect = DXGI_SWAP_EFFECT_SEQUENTIAL; // ORIGINAL = 0x1
 			//sd.SwapEffect = DXGI_SWAP_EFFECT_DISCARD;
-			//sd.SwapEffect = DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL;
+			//sd.SwapEffect = DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL; // Recommended for better performance. Can we use it?
```
I _think_ I tried this, but didn't notice much difference in performance. Yes, I think we can try this for OpenXR

morallo replied:
I have enabled it, it does not seem to have any unwanted side effect, and it enables for example overlays (FPS)

morallo committed:
Create an actual instance of the VRRendererOpenXR class

morallo committed:
Create an actual instance of the VRRendererOpenXR class

morallo committed:
OpenXR frame synchronization and layer rendering.

Currently it only displays a black screen both in the HMD and in the 2D monitor.
Headtracking still not implemented.

morallo committed:
OpenXR frame synchronization and layer rendering.

Currently it only displays a black screen both in the HMD and in the 2D monitor.
Headtracking still not implemented.

morallo committed:
MessageBox with OpenXR runtime error

morallo committed:
MessageBox with OpenXR runtime error

morallo committed:
WIP initialization of eye transform matrices

morallo committed:
WIP initialization of eye transform matrices

morallo committed:
1st attempt at getting the right eye projections

morallo committed:
1st attempt at getting the right eye projections

morallo committed:
Add debug output to troubleshoot EndFrame()

The ID3D11Device is being removed when EndFrame is called. It returns XR_ERROR_HANDLE_INVALID.

morallo committed:
Add debug output to troubleshoot EndFrame()

The ID3D11Device is being removed when EndFrame is called. It returns XR_ERROR_HANDLE_INVALID.

morallo committed:
Add flag to signal unfinished frames for sync

morallo committed:
Add flag to signal unfinished frames for sync

Prof-Butts committed:
Debug code for OpenXR. Remove Matrix code override later.

Prof-Butts committed:
Debug code for OpenXR. Remove Matrix code override later.

morallo committed:
Lots of debug to troubleshoot D3D Device removed

morallo committed:
Lots of debug to troubleshoot D3D Device removed

morallo committed:
Fixed buffer copy to HMD

The OpenXR swapchains were being created associated to the wrong D3DDevice/context due to DeviceResources::Initialize() being called twice. This caused a D3DDevice removal when trying to copy between different contexts.

morallo committed:
Fixed buffer copy to HMD

The OpenXR swapchains were being created associated to the wrong D3DDevice/context due to DeviceResources::Initialize() being called twice. This caused a D3DDevice removal when trying to copy between different contexts.

morallo committed:
Fixed projection matrix from OpenXR.

The conversion from XMFLOAT4X4 to Matrix4 needed to be transposed.
There is still an issue with the asymmetric FOV apparently. The 2D menu image is cropped on the small FOV side (nose side).

morallo committed:
Fixed projection matrix from OpenXR.

The conversion from XMFLOAT4X4 to Matrix4 needed to be transposed.
There is still an issue with the asymmetric FOV apparently. The 2D menu image is cropped on the small FOV side (nose side).

morallo committed:
FOV from OpenXR, fully disable BentBuf

Apply the FOV provided by OpenXR by setting g_fVR_FOV once the FOV is calculated during the first call to StereoRenderer::UpdateViewMatrices().

Fully disable Bent normal buffers, since the leftover code caused a crash when switching to 3D mode.

morallo committed:
FOV from OpenXR, fully disable BentBuf

Apply the FOV provided by OpenXR by setting g_fVR_FOV once the FOV is calculated during the first call to StereoRenderer::UpdateViewMatrices().

Fully disable Bent normal buffers, since the leftover code caused a crash when switching to 3D mode.

Prof-Butts committed:
Added Gamma Fix shaders, buffers and RTVs

This commits adds an extra (optional) step that converts the final buffers from
UNORM to SRGB. To enable it, add the following line to DDraw.cfg:

EnableGammaFix = 1

Prof-Butts committed:
Added Gamma Fix shaders, buffers and RTVs

This commits adds an extra (optional) step that converts the final buffers from
UNORM to SRGB. To enable it, add the following line to DDraw.cfg:

EnableGammaFix = 1

morallo committed:
OpenXR headtracking for 2D and fix gamma in HMD

morallo committed:
OpenXR headtracking for 2D and fix gamma in HMD

morallo committed:
First attempt at DLL export

morallo committed:
Rebase into master

morallo committed:
WIP merge
